### ISSUE TITLE
Support @mel.unwrap with uncurry

### ISSUE THREAD

**Issue by tatchi** [melange-re/melange#1027]
I was trying to create bindings for the following Js code:

```js
let navigate = useNavigate();

// Then later on in e.g an event handler:
navigate("/path") // or navigate(-1)
```

I tried the following reason code

```reason
[@mel.module "react-router-dom"]
external useNavigate:
  (unit, [@mel.unwrap] [ | `Str(string) | `Int(int)]) => unit =
  "useNavigate";

let navigate = useNavigate();

navigate(`Int(-1))
```

But it doesn't compile to what I need (related to https://github.com/melange-re/melange/issues/1022)

```js
import * as ReactRouterDom from "react-router-dom";

function navigate(param) {
  ReactRouterDom.useNavigate(param.VAL);
}

ReactRouterDom.useNavigate(-1);

export {
  navigate ,
}
```

I thought I would use an [uncurried version](https://melange.re/unstable/playground/?language=Reason&code=W0BtZWwubW9kdWxlICJyZWFjdC1yb3V0ZXItZG9tIl0KZXh0ZXJuYWwgdXNlTmF2aWdhdGU6CiAgdW5pdCA9PiAoLiAoW0BtZWwudW53cmFwXSBbIHwgYFN0cihzdHJpbmcpIHwgYEludChpbnQpXSkpID0%2BIHVuaXQgPQogICJ1c2VOYXZpZ2F0ZSI7CgpsZXQgbmF2aWdhdGUgPSB1c2VOYXZpZ2F0ZSgpOwpuYXZpZ2F0ZSguIGBTdHIoImhlbGxvIikpOwo%3D&live=off) but it doesn't work with `@mel.unwrap`

```reason
[@mel.module "react-router-dom"]
external useNavigate:
  unit => (. ([@mel.unwrap] [ | `Str(string) | `Int(int)])) => unit =
  "useNavigate";

let navigate = useNavigate();
navigate(. `Str("hello"));
```

It gives the following alert:

```
Line 3, 16:
Alert: unused Unused attribute [@mel.unwrap]
This means such annotation is not annotated properly.
For example, some annotations are only meaningful in externals
```

The uncurried version without `@mel.unwrap` compiles to what I'd need (except that the argument is not unwrapped ofc)

```reason
[@mel.module "react-router-dom"]
external useNavigate: unit => (. [ | `Str(string) | `Int(int)]) => unit =
  "useNavigate";

let navigate = useNavigate();
navigate(. `Str("/path"));
```

```js
import * as ReactRouterDom from "react-router-dom";

var navigate = ReactRouterDom.useNavigate();

navigate({
      NAME: "Str",
      VAL: "/path"
    });

export {
  navigate ,
}
```

Should we support `@mel.unwrap` with uncurry?

**Comment by anmonteiro** [comment]
In the meantime you can try Js.undefined instead of `unit` if you really need the undefined argument.

**Comment by tatchi** [comment]
Also note that in rescript 11, we can use `unboxed` with multiple constructors, so the following code compile to what I need:

```rescript
@unboxed
type stringOrInt = Str(string) | Int(int)

@module("react-router-dom")
external useNavigate: unit => stringOrInt => unit = "useNavigate"

let navigate = useNavigate()

navigate(Str("/path"))
```

Would supporting `unboxed` for multiple constructors be an option in melange?

**Comment by tatchi** [comment]
> In the meantime you can try Js.undefined instead of `unit` if you really need the undefined argument.

Thanks, I tried with

```reason
[@mel.module "react-router-dom"]
external useNavigate:
  (Js.undefined(unit), [@mel.unwrap] [ | `Str(string) | `Int(int)]) => unit =
  "useNavigate";

let navigate = useNavigate(Js.undefined);

navigate(`Int(-1));
```

But it stills doesn't compile to what I need:

```js
import * as ReactRouterDom from "react-router-dom";

function navigate(param) {
  ReactRouterDom.useNavigate(undefined, param.VAL);
}

ReactRouterDom.useNavigate(undefined, -1);

export {
  navigate ,
}
```

Is it what you had in mind ?

**Comment by anmonteiro** [comment]
This exposes a wider issue that we don't support these attributes inside uncurried externals, e.g.:

```ocaml
external x: (. ([@mel.as "hello"] _)) => unit = "asd";
```

**Comment by anmonteiro** [comment]
fixed in #1140
